source/Ship.cpp

```diff
@@ -1751,7 +1751,28 @@ void Ship::Launch(list<shared_ptr<Ship>> &ships, vector<Visual> &visuals)
 			// Resupply any ships launching of their own accord.
 			if(!ejecting)
 			{
-				// TODO: Restock fighter weaponry that needs ammo.
+				// Determine which of the fighter's weapons we can restock.
+				auto toRefill = set<const Outfit *>();
+				for(const auto &hardpoint : bay.ship->Weapons())
+				{
+					const Weapon *weapon = hardpoint.GetOutfit();
+					const Outfit *ammo = weapon ? weapon->Ammo() : nullptr;
+					if(ammo && OutfitCount(ammo))
+						toRefill.insert(ammo);
+				}
+				// Transfer as much ammo as the fighter needs.
+				bool tookAmmo = false;
```

Comment:
How about *both*? Pasting in some code from your cargo version, this

1. Draws ammo from cargo
2. If cargo is insufficient, draw remaining ammo from outfits

I agree with petervdmeer's [comments on the other PR](https://github.com/endless-sky/endless-sky/pull/4127/files#r362398112), so I haven't touched the restock.  This would, however, in a roundabout way, allow a carrier to slowly replenish its own launchers from cargo as long as its matchingly armed fighters survive.

```suggestion
				// Transfer as much ammo as the fighter needs from cargo.
				bool tookAmmo = false;
				for(const Outfit *ammo : toRefill)
				{
					int canTake = bay.ship->attributes.CanAdd(*ammo, cargo.Get(ammo));
					if(canTake > 0)
					{
						cargo.Add(ammo, -canTake);
						bay.ship->AddOutfit(ammo, canTake);
						tookAmmo = true;
						carriedMass += ammo->Mass() * canTake;
					}
				}
				// Recalculate ammo needs.
				for(const auto &hardpoint : bay.ship->Weapons())
				{
					const Weapon *weapon = hardpoint.GetOutfit();
					const Outfit *ammo = weapon ? weapon->Ammo() : nullptr;
					if(ammo && OutfitCount(ammo))
						toRefill.insert(ammo);
				}
				// Transfer as much ammo as the fighter needs from outfits.
```

Reply:
I've basically done this, though I add a bit of flavor by restricting cargo restock to internally carried ships.

Reply:
Sorry, I should have made my comment about adding this in a separate PR here in this thread, instead of as a separate comment.